test/jdk/tools/jar/JarExtractTest.java

```diff
@@ -0,0 +1,382 @@
+        final Path dirOutsideScratchDir = Files.createTempDirectory(Paths.get(".."), "8173970");
+        // we need to explicitly delete this dir after the tests end
+        filesToDelete.add(dirOutsideScratchDir);
+        final String existing3 = dirOutsideScratchDir.toString() + File.separator + "8173970-existing-3";
+        Files.createDirectories(Paths.get(existing3));
+
+        final String anotherDirOutsideScratchDir = ".." + File.separator + "8173970-non-existent";
+        filesToDelete.add(Paths.get(anotherDirOutsideScratchDir));
+
+        return new Object[][]{
+                {"."}, // current dir
+                {"." + File.separator + "8173970-extract-1"}, // (explicitly) relative to current dir
+                {"8173970-extract-2"}, // (implicitly) relative to current dir
+                {anotherDirOutsideScratchDir}, // sibling to current dir
+                // some existing dirs
+                {existing1},
+                {existing2},
+                {existing3},
+                // a non-existent dir within an existing dir
+                {existing1 + File.separator + "non-existing" + File.separator + "foo"}
+        };
+    }
+
+    @DataProvider(name = "absExtractLocations")
+    private Object[][] provideAbsoluteExtractLocations() throws Exception {
+        final Object[][] relative = provideRelativeExtractLocations();
+        final Object[][] abs = new Object[relative.length][1];
+        int i = 0;
+        for (final Object[] p : relative) {
+            abs[i++][0] = Paths.get((String) p[0]).toAbsolutePath().toString();
+        }
+        return abs;
+    }
+
+    @DataProvider(name = "absNormalizedExtractLocations")
+    private Object[][] provideAbsoluteNormalizedExtractLocations() throws Exception {
+        final Object[][] relative = provideAbsoluteExtractLocations();
+        final Object[][] abs = new Object[relative.length][1];
+        int i = 0;
+        for (final Object[] p : relative) {
+            abs[i++][0] = Paths.get((String) p[0]).toAbsolutePath().normalize().toString();
+        }
+        return abs;
+    }
+
+    @Test(dataProvider = "relExtractLocations")
+    public void testExtractToRelativeDir(final String dest) throws Exception {
+        testExtract(dest);
+    }
+
+    @Test(dataProvider = "absExtractLocations")
+    public void testExtractToAbsoluteDir(final String dest) throws Exception {
+        testExtract(dest);
+    }
+
+    @Test(dataProvider = "absNormalizedExtractLocations")
+    public void testExtractToAbsoluteNormalizedDir(final String dest) throws Exception {
+        testExtract(dest);
+    }
+
+    /**
+     * Test that {@code jar -x -f --dir} works as expected
+     */
+    @Test
+    public void testExtractLongForm() throws Exception {
+        final String dest = "foo-bar";
+        System.out.println("Extracting " + testJarPath + " to " + dest);
+        final int exitCode = JAR_TOOL.run(System.out, System.err, "-x", "-f", testJarPath.toString(),
```

Comment:
Perhaps add a DataProvider so you can test --extract as well?

Reply:
Not a data provider but the latest version of PR tests --extract as well